### PR TITLE
cmd/collision.cpp: Make it so that Nav 8 cannot be crashed into

### LINKED ISSUE #516: PWCU: Unexpectedly, Nav 8 in Troy is an object that can be collided with

### PR DESCRIPTION
**Pull request by stephengtuggy**
Thank you for submitting a pull request and becoming a contributor to the Vega Strike Core Engine.

Please answer the following:

Code Changes:
- [X] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation
- [ ] This is a documentation change only

Issues:
- Please list any related issues
- Fixes #516 for 0.8.x branch

Purpose:
- What is this pull request trying to do? Fix a bug where, with PWCU, Nav 8 is unexpectedly an object that you can crash into
- What release is this for? 0.8.x; 0.9.x pending
- Is there a project or milestone we should apply this to? 0.8.x

### CODE REVIEW

File: engine/src/cmd/collision.cpp [hunk 1]
```diff
@@ -292,15 +292,24 @@ void Collision::collide( Unit* unit1,
     // See commit a66bdcfa1e00bf039183603913567d48e52c7a8e method Unit->jumpReactToCollision for an example
     // I assume this is for "always open" jump gates that you pass through
 
+    collision1.apply_force = false;
```
Comment (royfalk) [inline engine/src/cmd/collision.cpp:295]: This should not be necessary. The variables are initialized in the header.

File: engine/src/cmd/collision.cpp [hunk 2]
```diff
@@ -292,15 +292,24 @@ void Collision::collide( Unit* unit1,
...
+    collision2.apply_force = false;
+    collision2.deal_damage = false;
+
     collision1.shouldApplyForceAndDealDamage(unit2);
     collision2.shouldApplyForceAndDealDamage(unit1);
 
     float elasticity = 0.8f;
-    collision1.applyForce(elasticity, collision2.mass, collision2.velocity);
-    collision2.applyForce(elasticity, collision1.mass, collision1.velocity);
+    if (collision1.apply_force && collision2.apply_force) {
```
Comment (royfalk) [inline engine/src/cmd/collision.cpp:304]: Also not needed. First line of applyForce function returns if false.

File: engine/src/cmd/collision.cpp [hunk 3]
```diff
@@ -292,15 +292,24 @@ void Collision::collide( Unit* unit1,
...
+        collision1.applyForce(elasticity, collision2.mass, collision2.velocity);
+        collision2.applyForce(elasticity, collision1.mass, collision1.velocity);
+    }
 
-    collision1.dealDamage(collision2);
-    collision2.dealDamage(collision1);
+    if (collision1.deal_damage && collision2.deal_damage) {
```
Comment (royfalk) [inline engine/src/cmd/collision.cpp:309]: Same here. First line of dealDamage checks the variable.
In short, I don't see what your changes actually do to fix the issue. Also, I don't see how the issue is caused based on this. Unless you come up with a better hypothesis, I think the root cause is the function isJumppoint. It should return without setting either apply_force or apply_damage to true.
As this appears to work, we should probably check the data or WC to see what's there.

File: engine/src/cmd/collision.cpp (hunk 1 above)
Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp:295]: You might be right. But it wasn't obvious from looking at the code whether these fields were initialized to `false`, or not.

File: engine/src/cmd/collision.cpp (hunk 2 above)
Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp:304]: The behavior definitely changed when I made this change to the code. The difference is that now, no force is applied to **either** object/unit unless **both** pass the "shouldApplyForce" test. Whereas before, the conditions applied to the units individually.

File: engine/src/cmd/collision.cpp (hunk 3 above)
Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp:309]: See previous comment, above.

Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp:309]: You might be right about isJumppoint though. I'll take a look at that as well.

**Review by stephengtuggy — COMMENTED** [review]
@royfalk See responses.

File: engine/src/cmd/collision.cpp (hunk 1 above)
Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp:295]: You're right. I see that now. I just took those redundant initializations out.

**Comment by stephengtuggy** [comment]
All right, let me work on this issue some more. I think I can find a better solution.

File: engine/src/cmd/collision.cpp (hunk 2 above)
Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp]: OK, I went ahead and undid this. It didn't turn out to be necessary after all.

File: engine/src/cmd/collision.cpp (hunk 3 above)
Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp]: Undid this too

**Comment by stephengtuggy** [comment]
OK, I think I like this solution better. To determine whether a planet can be collided with, It literally checks for the string `"invisible"` in the planet's `getFullname()` property. So actual planets -- like Atlantis or Troy -- can be collided with, while invisible `Nav` objects -- such as `Nav_8` cannot.

It's a bit of an ugly hack, but it works.

**Comment by stephengtuggy** [comment]
@royfalk you were correct -- none of my 3 earlier code changes were necessary. I commented out or removed each one of them, and took a different approach.

**Comment by stephengtuggy** [comment]
Probably the ultimate solution here would be to have each game / dataset define its own types of objects, such as planets, warp points, etc., and define which ones can be collided with and which ones can't.

File: engine/src/cmd/collision.cpp [hunk 4]
```diff
@@ -101,17 +101,26 @@ _UnitType::enhancement,
...
+            // Handle the "Nav 8" case
+            // BOOST_LOG_TRIVIAL(debug) << "shouldApplyForceAndDealDamage(): other_unit is a planet, with full name " << other_unit->getFullname();
+            if (other_unit->getFullname().find("invisible") != std::string::npos) {
+                // BOOST_LOG_TRIVIAL(debug) << "Found a Nav_8-type object";
+                return;
+            } else {
+                break;
+            }
+        default:
+            break;
```
Comment (BenjamenMeyer) [inline engine/src/cmd/collision.cpp:123]: should probably do a debug warning on default

File: engine/src/cmd/collision.cpp [hunk 5]
```diff
@@ -101,17 +101,26 @@ _UnitType::enhancement,
 _UnitType::missile*/
 void Collision::shouldApplyForceAndDealDamage(Unit* other_unit)
 {
-    // Collision with a nebula does nothing
-    if(other_unit->isUnit() == _UnitType::nebula)
+    switch (other_unit->isUnit())
```
Comment (BenjamenMeyer) [inline engine/src/cmd/collision.cpp:104]: switch over if-block is 100% better!

**Review by BenjamenMeyer — CHANGES_REQUESTED** [review]

**Comment by BenjamenMeyer** [comment]
@stephengtuggy  we probably just need to add another object type for the NAV Point object to the engine.

The data sets should be able to change the displayable name various types, but should just be using what is provided by the engine.

**Comment by stephengtuggy** [comment]
> @stephengtuggy we probably just need to add another object type for the NAV Point object to the engine.
> 
> The data sets should be able to change the displayable name various types, but should just be using what is provided by the engine.

@BenjamenMeyer do you want to add the Nav_Point object type in 0.8.x? I suppose we could do so in a fairly backwards-compatible manner, since we would be adding a unit type, not subtracting anything.

**Comment by BenjamenMeyer** [comment]
> > @stephengtuggy we probably just need to add another object type for the NAV Point object to the engine.
> > The data sets should be able to change the displayable name various types, but should just be using what is provided > @stephengtuggy we probably just need to add another object type for the NAV Point object to the engine.
> The data sets should be able to change the displayable name various types, but should just be using what is provided by the engine.

@BenjamenMeyer do you want to add the Nav_Point object type in 0.8.x? I suppose we could do so in a fairly backwards-compatible manner, since we would be adding a unit type, not subtracting anything.by the engine.
> 
> @BenjamenMeyer do you want to add the Nav_Point object type in 0.8.x? I suppose we could do so in a fairly backwards-compatible manner, since we would be adding a unit type, not subtracting anything.

if it's not hard to do then we may have the time to; but otherwise it could be 0.9.x.

**Comment by stephengtuggy** [comment]
@BenjamenMeyer so then what changes are required for 0.8.x? Just the warning message if a unit type does not match one of the expected values?

**Comment by BenjamenMeyer** [comment]
> @BenjamenMeyer so then what changes are required for 0.8.x? Just the warning message if a unit type does not match one of the expected values?

At a minimum yes; but let's see how much effort it would be to add a NavPoint object that could be utilized. If it's too much work, we'll push that to 0.9.x; if not, it can go in 0.8.x but let's see what it would take first.

File: engine/src/cmd/collision.cpp [hunk 6]
```diff
@@ -101,17 +101,26 @@ _UnitType::enhancement,
...
+            // Collision with a nebula does nothing
+            return;
+        case _UnitType::enhancement:
+            // Collision with a enhancement improves your shield apparently
+            apply_force = true;
+            return;
+        case _UnitType::planet:
+            // Handle the "Nav 8" case
+            // BOOST_LOG_TRIVIAL(debug) << "shouldApplyForceAndDealDamage(): other_unit is a planet, with full name " << other_unit->getFullname();
+            if (other_unit->getFullname().find("invisible") != std::string::npos) {
```
Comment (royfalk) [inline engine/src/cmd/collision.cpp:116]: It's an ugly hack, but at least it's minimal.

**Comment by royfalk** [comment]
I was looking for an approve button when I realized there isn't one. LGTM.

**Comment by stephengtuggy** [comment]
As I'm thinking some more about this, it occurs to me that if we add a NavPoint object to 0.8.x, then the game engine will be backwards compatible, but the PWCU data set won't be. That may not be desirable. I think that for the 0.8.x timeframe, we should keep both (all?) game datasets compatible with previous versions of the Vega Strike Game Engine. Which would mean leaving the NavPoint and similar objects for later.

What do you all think?

File: engine/src/cmd/collision.cpp (hunk 6 above)
Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp:116]: @royfalk Yep.

File: engine/src/cmd/collision.cpp (hunk 4 above)
Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp:132]: Done. For both the current unit's type and the other unit's type, I now handle all possible values of the `_UnitType` enum, and warn if an unexpected value is encountered.

**Comment by stephengtuggy** [comment]
Play tests rerun, with both PWCU and VS:UtCS, and on Windows as well as Linux. Seems to be working well.

**Comment by stephengtuggy** [comment]
The more I think about it, the more complicated I think it will be to add a NavPoint unit type to the code. I vote we wait until 0.9.x to do that.

**Comment by stephengtuggy** [comment]
> The more I think about it, the more complicated I think it will be to add a NavPoint unit type to the code. I vote we wait until 0.9.x to do that.

We will need to add some more possible values to the `_UnitType` enum. Add code to handle the additional values in each place where it's relevant. Add classes for the new unit types. Implement physics and drawing behavior. Then use the new functionality in the PWCU dataset. And test, test, test.

**Review by BenjamenMeyer — APPROVED** [review]

**Comment by BenjamenMeyer** [comment]
@stephengtuggy thanks for the analysis. We can let it slide for 0.8.x

**Review by Loki1950 — APPROVED** [review]

**Pull request merged**